Introduce operator aliases

garyb opened the pull request:
Resolves #806

This probably isn't the best way of handling it in the long term, it would be nice to desugar during codegen rather than upfront as we'll probably get better error messages that way, but while we're supporting both old and new syntax this seemed a reasonable approach for now.

garyb commented:
I've done some testing with this too - made a local version of the prelude that has all its operators defined as aliases and then compiled a Halogen example that still works succesfully. Also checked the output is correct in the presence of re-exports of aliases.

Once this and my other outstanding PRs are in I'd like to revisit purescript/purescript-prelude#41, update the compiler for any changed identifier locations, and then release a prelude 1.0.0 to go along with the psc 0.8 release, if that sounds good?

paf31 commented on src/Language/PureScript/Docs/Types.hs:
```diff
@@ -307,9 +307,12 @@ asDeclaration =
+asFixity = do
+  fixity <- P.Fixity <$> key "associativity" asAssociativity
+                     <*> key "precedence" asIntegral
+  alias <- key "alias" (perhaps asString)
```
So this needs an update to Pursuit, correct? Or is this a backwards-compatible change? Does `perhaps` handle undefined fields?

Maybe we need to bump the min compiler version there. /cc @hdgarrood

garyb replied:
I'm not sure, so yeah, one for @hdgarrood.

hdgarrood replied:
Apparently not, `perhaps` is kind of like purescript's `Data.Foreign.Null.readNull`: https://hackage.haskell.org/package/aeson-better-errors-0.9.0/docs/Data-Aeson-BetterErrors.html#v:perhaps
But there is also `keyMay`, which would give you `Nothing` if the property wasn't there, so as far as I can tell, if we used that, we would probably be able to keep accepting uploads from people on 0.7.x. I'm not really sure about what to do here; it's hard to predict whether that would be worth doing.

By the way, does this line up with the `ToJSON` instance? You can run the test in `psc-publish/test` to do a quick to-JSON-and-back test. (I should make that into an actual test.) I think this might be using Aeson's `ToJSON (,)` instance at the moment, which might not work with this parser.

paf31 commented on src/Language/PureScript/Parser/Declarations.hs:
```diff
@@ -124,8 +124,9 @@ parseFixityDeclaration :: TokenParser Declaration
 parseFixityDeclaration = do
   fixity <- parseFixity
   indented
+  alias <- P.optionMaybe $ parseIdent <* reserved "as"
```
`parseIdent` -> `lname`? We don't want operators here, right?

garyb replied:
Ah, good point!

paf31 commented on src/Language/PureScript/Sugar/Operators.hs:
```diff
@@ -1,22 +1,3 @@
--- Portability :
---
--- |
--- This module implements the desugaring pass which reapplies binary operators based
```
Worth keeping the comments?

garyb replied:
Oops, yep. I did so with the other ones I took the header out of but missed it here.

paf31 commented on src/Language/PureScript/Sugar/Operators.hs:
```diff
@@ -57,9 +40,24 @@ import qualified Language.PureScript.Constants as C
+  renameAliasedOperators aliased (Module ss coms mn ds exts) = Module ss coms mn (map f' ds) exts
+    where
+    (f', _, _) = everywhereOnValues id go id
+    go (Var name) = Var $ fromMaybe name (name `M.lookup` aliased)
```
If the value is not in the map, should that be an `internalError`?

garyb replied:
No, because this matches all `Var`s, the map only has the aliased operators in it, so if the lookup fails it just means it's not an operator, or one that was defined the old way.

paf31 replied:
Oh right. But if we matched `Op` only, then in 0.9 it could be an error?

garyb replied:
Yeah exactly :) There are quite a few details of this that can be simplified in 0.9 I think.

paf31 commented:
I like it! I tried this before but didn't think of the renaming-after-reordering approach for some reason, which threw me off. 

Agreed about handling this in codegen, but this is great for now.

Prelude changes sound good to me, but why 1.0? Don't we want to wait until other things are ready and bump everything together?

garyb commented:
Updated - `identifier` is the one we wanted rather than `lname` I think, as that covers `_`-prefixed idents too. Added the comments back to the operators sugar module.

It would seem that using `perhaps` in the aeson stuff will probably be backwards compatible [given the description here](https://hackage.haskell.org/package/aeson-better-errors-0.9.0/docs/src/Data-Aeson-BetterErrors-Internal.html#perhaps), but Harry should definitely know given it's his library :smile:

hdgarrood commented on src/Language/PureScript/Docs/Convert.hs:
```diff
@@ -97,7 +100,7 @@ augmentDeclarations (partitionEithers -> (augments, toplevels)) =
 addDefaultFixity decl@Declaration{..}
   | isOp declTitle && isNothing declFixity =
-        decl { declFixity = Just defaultFixity }
+        decl { declFixity = Just (defaultFixity, Nothing) }
```
Just checking: once we require operators to be defined as aliases for other existing values, this `addDefaultFixity` function will no longer be necessary, right?

garyb replied:
Yeah, that's right.

garyb commented:
Switched to use `keyMay` instead of `perhaps`. I couldn't get the `psc-publish` tests to work though, so I'm not sure if the mentioned JSON instance issue is a problem or not.

hdgarrood commented:
Ok - I'll try to fix those tests up and integrate them into the normal test suite.

hdgarrood commented:
@garyb can you cherry-pick https://github.com/hdgarrood/purescript/commit/4ab687c69de1b6ea3695c31ded47bb1663cc7bb5?

garyb commented:
@hdgarrood done! (A while back actually, Travis is pretty slow currently). Weirdly failing in just one of the test setups...

hdgarrood commented:
Ah, looks like that new module `TestPscPublish` needs to be added to the `other-modules` section in the cabal file. Although I have no idea why it only failed once...

phadej commented:
@hdgarrood because `RUNSDISTTESTS=YES` is only in one travis job. There are almost no sense to run it with all compiler versions, would just make builds slower.

hdgarrood commented:
Ah, of course - thanks.

garyb commented:
That did it, thanks. I think this is ready to go if you're happy then @paf31?

paf31 commented:
:+1: Looks great!

garyb commented:
:confetti_ball:
